src/baseline.rs

```diff
@@ -7,22 +7,232 @@ use crate::dump::RustDocCommand;
 use crate::util::slugify;
 use crate::GlobalConfig;
 
+mod crate_features {
+    /// Sometimes crates ship types with fields or variants that are included
+    /// only when certain features are enabled.
+    ///
+    /// By default, we want to generate rustdoc with `--all-features`,
+    /// but that option isn't available outside of the current crate,
+    /// so we have to implement it ourselves.
+
+    pub(crate) fn get_all_crate_features_from_registry(
+        crate_: &crates_index::Version,
+    ) -> Vec<String> {
+        // Implicit features from optional dependencies have to be added separately
+        // from regular features: https://github.com/obi1kenobi/cargo-semver-checks/issues/265
+        let mut implicit_features: std::collections::BTreeSet<_> = crate_
+            .dependencies()
+            .iter()
+            .filter_map(|dep| dep.is_optional().then_some(dep.name()))
+            .map(|x| x.to_string())
+            .collect();
+        for feature_defn in crate_.features().values().flatten() {
+            // "If you specify the optional dependency with the dep: prefix anywhere
+            //  in the [features] table, that disables the implicit feature."
+            // https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies
+            if let Some(optional_dep) = feature_defn.strip_prefix("dep:") {
+                implicit_features.remove(optional_dep);
+            }
+        }
+        let regular_features: std::collections::BTreeSet<_> =
+            crate_.features().keys().cloned().collect();
+        let mut all_crate_features = implicit_features;
+        all_crate_features.extend(regular_features);
+        all_crate_features.into_iter().collect()
+    }
+
+    #[allow(unused_variables)]
+    pub(crate) fn get_all_crate_features_from_manifest(path: &std::path::Path) -> Vec<String> {
+        unimplemented!()
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+enum CrateSource<'a> {
+    Registry { crate_: &'a crates_index::Version },
+    ManifestPath { path: &'a Path, name: String },
+}
+
+impl<'a> CrateSource<'a> {
+    fn get_name(&self) -> String {
+        match self {
+            Self::Registry { crate_ } => crate_.name().to_string(),
+            Self::ManifestPath { .. } => unimplemented!(),
+        }
+    }
+
+    fn get_version(&self) -> String {
+        match self {
+            Self::Registry { crate_ } => crate_.version().to_string(),
+            Self::ManifestPath { .. } => unimplemented!(),
+        }
+    }
```

Comment:
Two suggestions:
- The underlying code uses `.name()` and `.version()` for these accessor methods, so let's match that naming convention for simplicity.
- Let's return a reference and avoid allocating, since not all uses of this function will care about owning the result, and ones that do are welcome to do `.to_string()` themselves. This preserves maximum flexibility for the caller and again matches the upstream API.
```suggestion
    fn name(&self) -> &str {
        match self {
            Self::Registry { crate_ } => crate_.name(),
            Self::ManifestPath { .. } => unimplemented!(),
        }
    }

    fn version(&self) -> &str {
        match self {
            Self::Registry { crate_ } => crate_.version(),
            Self::ManifestPath { .. } => unimplemented!(),
        }
    }
```

Reply:
I see. So the convention is to by default return `&str` instead of `String`?

Reply:
Generally, it's better to return types that assume less and enable more for the caller. The caller can easily make `&str` into a `String` if they want to, but if we return `String` here, they can't retroactively opt out of the allocation and copy.

Reply:
Done in d284a1692ae9885ab16602bb66c8f0de14e76819 and 91114184ecadea4331ffbb920381507ec2ee0c45.